Sources/SwiftCoAP/SCMessage.swift

```diff
@@ -32,87 +34,171 @@ protocol SCCoAPTransportLayerDelegate: class {
+                        return
+                    }
+                    if let data = data {
+                        self.transportLayerDelegate.transportLayerObject(self, didReceiveData: data, fromHost: hostPort.host, port: hostPort.port)
+                    }
+                }
+            }
         }
-        return true
     }
 }
 
 extension SCCoAPUDPTransportLayer: SCCoAPTransportLayerProtocol {
-    func sendCoAPData(_ data: Data, toHost host: String, port: UInt16) throws {
-        try startListening()
-        udpSocket.send(data, toHost: host, port: port, withTimeout: 0, tag: udpSocketTag)
-        udpSocketTag = (udpSocketTag % Int.max) + 1
-    }
-    
-    func closeTransmission() {
-        udpSocket.close()
-        udpSocket = nil
+    public func sendCoAPData(_ data: Data, toHost host: String, port: UInt16) throws {
+        var connection = mustGetConnection(forHost: host, port: port)
+        let hostPort = HostPortKey(host: host, port: port)
+        if connection.stateUpdateHandler == nil {
+            connection = setupStateUpdateHandler(for: connection, withHostPort: hostPort)
+        }
+        if connection.state == .setup {
+            connection.start(queue: DispatchQueue.global(qos: .utility))
+        }
+        connection.send(content: data, completion: .contentProcessed{ [weak self] error in
+            guard let self = self else { return }
+            if error != nil {
+                self.transportLayerDelegate?.transportLayerObject(self, didFailWithError: error! as NSError)
+            }
+        })
     }
     
-    func startListening() throws {
-        if udpSocket == nil && !setUpUdpSocket() {
-            udpSocket.close()
-            udpSocket = nil
-            throw SCCoAPTransportLayerError.setupError(errorDescription: "Failed to setup UDP socket")
+    public func closeTransmission() {
+        connections.forEach{
+            $0.value.cancel()
         }
-    }
-}
-
-extension SCCoAPUDPTransportLayer: GCDAsyncUdpSocketDelegate {
-    func udpSocket(_ sock: GCDAsyncUdpSocket!, didReceive data: Data!, fromAddress address: Data!, withFilterContext filterContext: Any!) {
-        transportLayerDelegate.transportLayerObject(self, didReceiveData: data, fromHost: GCDAsyncUdpSocket.host(fromAddress: address), port: GCDAsyncUdpSocket.port(fromAddress: address))
+        connections = [:]
     }
     
-    func udpSocket(_ sock: GCDAsyncUdpSocket!, didNotSendDataWithTag tag: Int, dueToError error: Error!) {
-        transportLayerDelegate.transportLayerObject(self, didFailWithError: error as NSError)
+    public func startListening() throws {
+        listener = try NWListener(using: .udp, on: 5683)
+        listener?.newConnectionHandler = { [weak self] newConnection in
+            guard let self = self else { return }
+            var connection: NWConnection!
+            var hostPort: HostPortKey!
```

Comment:
```suggestion
            let hostPort: HostPortKey
```
Isn't that equivalent and safer?